### PR TITLE
1.19.4

### PR DESCRIPTION
**Pull request by MelonHell**

### CODE REVIEW

**Comment by ZakShearman** [comment]
Is this intended to be working? or is it a draft?

**Comment by MelonHell** [comment]
> Is this intended to be working? or is it a draft?

it works but needs more testing

File: src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        return super.metadata.getIndex(OFFSET + 3, Vec.ONE);
+    }
+
+    public void setScale(@NotNull Vec value) {
+        super.metadata.setIndex(OFFSET + 3, Metadata.Vector3(value));
+    }
+
+    public float @NotNull[] getRightRotation() {
+        //todo replace with actual quaternion type
+        return super.metadata.getIndex(OFFSET + 4, new float[] {0, 0, 0, 1});
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java:53]: replace with quaternion float array

File: src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public void setRightRotation(float @NotNull[] value) {
+        super.metadata.setIndex(OFFSET + 4, Metadata.Quaternion(value));
+    }
+
+    public float @NotNull[] getLeftRotation() {
+        //todo replace with actual quaternion type
+        return super.metadata.getIndex(OFFSET + 5, new float[] {0, 0, 0, 1});
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java:62]: replace with quaternion float array

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    public int getLineWidth() {
+        return super.metadata.getIndex(OFFSET + 1, 200);
+    }
+
+    public void setLineWidth(int value) {
+        super.metadata.setIndex(OFFSET + 1, Metadata.VarInt(value));
+    }
+
+    public int getBackgroundColor() {
+        return super.metadata.getIndex(OFFSET + 2, 1073741824);
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:39]: replace with background colour

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    public @NotNull Component getText() {
+        return super.metadata.getIndex(OFFSET, Component.empty());
+    }
+
+    public void setText(@NotNull Component value) {
+        super.metadata.setIndex(OFFSET, Metadata.Chat(value));
+    }
+
+    public int getLineWidth() {
+        return super.metadata.getIndex(OFFSET + 1, 200);
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:31]: replace with line width

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    public int getBackgroundColor() {
+        return super.metadata.getIndex(OFFSET + 2, 1073741824);
+    }
+
+    public void setBackgroundColor(int value) {
+        super.metadata.setIndex(OFFSET + 2, Metadata.VarInt(value));
+    }
+
+    public int getTextOpacity() {
+        return super.metadata.getIndex(OFFSET + 3, -1);
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:47]: replace with opacity

File: src/main/java/net/minestom/server/entity/metadata/other/InteractionMeta.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    public float getWidth() {
+        return super.metadata.getIndex(OFFSET, 1f);
+    }
+
+    public void setWidth(float value) {
+        super.metadata.setIndex(OFFSET, Metadata.Float(value));
+    }
+
+    public float getHeight() {
+        return super.metadata.getIndex(OFFSET + 1, 1f);
```
Comment (iam4722202468) [inline src/main/java/net/minestom/server/entity/metadata/other/InteractionMeta.java:26]: replace with height

**Review by iam4722202468 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java (hunk 4 above)
Comment (MelonHell) [reply src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:31]: the problem is the data type or the default value?

**Comment by MelonHell** [comment]
I'm thinking about how to deal with ARGB colors in Minestom, I need to come up with something to add transparency support, but without breaking backward compatibility

**Comment by Codestech1** [comment]
On merge i'll start making 1.20 pr 😎

**Comment by MelonHell** [comment]
I need help with ARGB colors. It seems to me that ARGB color should not be a separate class, but I do not know how to add an alpha channel to Color without breaking backwards compatibility

File: src/main/java/net/minestom/server/entity/Metadata.java [hunk 7]
```diff
@@ -31,7 +32,7 @@ public static Entry<Integer> VarInt(int value) {
     }
 
     public static Entry<Long> Long(long value) {
-        return new MetadataImpl.EntryImpl<>(TYPE_LONG, value, NetworkBuffer.LONG);
+        return new MetadataImpl.EntryImpl<>(TYPE_LONG, value, NetworkBuffer.VAR_LONG);
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/entity/Metadata.java:35]: This has overflowed from 1.19.3 but `#Long` should be renamed to `#VarLong` and `TYPE_LONG` to `TYPE_VAR_LONG`

File: src/main/java/net/minestom/server/entity/metadata/animal/CamelMeta.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package net.minestom.server.entity.metadata.animal;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.entity.Metadata;
+import org.jetbrains.annotations.NotNull;
+
+public class CamelMeta extends AbstractHorseMeta {
+    public static final byte OFFSET = AbstractHorseMeta.MAX_OFFSET;
+    public static final byte MAX_OFFSET = OFFSET + 1;
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/entity/metadata/animal/CamelMeta.java:9]: Pretty sure this is supposed to be 1 above the max although I'm not sure as it appears pretty inconsistent.

E.g:
  - BatMeta has 1 field and is +0 (same concept as this)
  - AgeableMobMeta has 1 field and is +1 (so 1 above the max index it uses)
  - FireballMeta has 0 fields and is +0 (so same theory as max + 1, otherwise it would have to be `MAX_OFFSET = OFFSET - 1` which seems obviously wrong)

Basically I have no idea what this should be. Maybe a separate PR later to make them all consistent.

**Review by ZakShearman — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/entity/metadata/animal/CamelMeta.java (hunk 8 above)
Comment (ZakShearman) [reply src/main/java/net/minestom/server/entity/metadata/animal/CamelMeta.java:9]: Confirmed, it should be 1 above the max so `public static final byte MAX_OFFSET = OFFSET + 2;`

File: src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java
[the remark this answers is not in the document]
Comment (MelonHell) [reply src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java:1]: image
I also find this strange. I didn't quite understand what @iam4722202468 wanted

File: src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import net.minestom.server.network.packet.server.ServerPacket;
+import net.minestom.server.network.packet.server.ServerPacketIdentifier;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static net.minestom.server.network.NetworkBuffer.*;
+
+// Notes
+// sourceEntityId - 0 indicates no source entity, otherwise it is entityId + 1
+// sourceDirectId - 0 indicates no direct source. Direct attacks (e.g. melee) will have this number me the same as sourceEntityId, indirect attacks (e.g. projectiles) will have this be be the projectile entity id + 1
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java:14]: Might be worth expanding on the javadocs for this as below. Please do proof read:

```suggestion
/**
 * See <a href="https://wiki.vg/Protocol#Damage_Event">https://wiki.vg/Protocol#Damage_Event</a> for more info.
 *
 * @param targetEntityId ID of the entity being damaged
 * @param damageTypeId   0 if there is no damage type, otherwise the damage type ID + 1
 * @param sourceEntityId 0 if there is no source entity, otherwise it is entityId + 1
 * @param sourceDirectId 0 if there is no direct source. For direct attacks (e.g. melee), this is the same as sourceEntityId. For indirect attacks (e.g. projectiles), this is the projectile entity id + 1
 * @param sourcePos      null if there is no source position, otherwise the position of the source
 */
```

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            if (rightValue) {
+                mask |= ALIGN_RIGHT;
+            } else {
+                mask &= ~ALIGN_RIGHT;
+            }
+        }
+        setMask(index, mask);
+    }
+
+    public enum TextAlignment{
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:113]: Missing space between enum name and `{`

File: src/main/java/net/minestom/server/registry/Registry.java [hunk 11]
```diff
@@ -142,6 +142,7 @@ public enum Resource {
         POTION_EFFECTS("potion_effects.json"),
         POTION_TYPES("potions.json"),
         PARTICLES("particles.json"),
+        DAMAGE_TYPES("damage_types.json"),
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/registry/Registry.java:145]: Damage types need to be autogenerated into Java

File: src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    public int getId() {
+        return ServerPacketIdentifier.DAMAGE_EVENT;
+    }
+
+    @Override
+    public void write(@NotNull NetworkBuffer writer) {
+        writer.write(VAR_INT, targetEntityId);
+        writer.write(VAR_INT, targetEntityId);
+        writer.write(VAR_INT, targetEntityId);
+        writer.write(VAR_INT, targetEntityId);
```
Comment (ZakShearman) [inline src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java:32]:
```suggestion
    public void write(@NotNull NetworkBuffer writer) {
        writer.write(VAR_INT, targetEntityId);
        writer.write(VAR_INT, damageTypeId);
        writer.write(VAR_INT, sourceEntityId);
        writer.write(VAR_INT, sourceDirectId);
```

**Review by ZakShearman — CHANGES_REQUESTED** [review]

File: src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java (hunk 4 above)
Comment (iam4722202468) [reply src/main/java/net/minestom/server/entity/metadata/display/TextDisplayMeta.java:32]: I was mistaken, this is fine

File: src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java
[the remark this answers is not in the document]
Comment (iam4722202468) [reply src/main/java/net/minestom/server/entity/metadata/display/AbstractDisplayMeta.java:1]: I was looking over the PR quickly and misread the code, what you were doing was fine. The method should indeed be called getHeight, as this makes sense and matches all other metadata getters.

File: src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java (hunk 12 above)
Comment (MelonHell) [reply src/main/java/net/minestom/server/network/packet/server/play/DamageEventPacket.java]: lol how did i not notice that

**Comment by MelonHell** [comment]
I wrote an implementation of the new DamageType, but I don't understand what to do with the old net.minestom.server.entity.damage.DamageType

I also doubt my code, please check and correct

**Comment by ZakShearman** [comment]
This change should be added https://github.com/hollow-cube/minestom-ce/commit/c8358d64a176bdf7ef2c51f0cb1f5c05d0fa2543

**Comment by Codestech1** [comment]
> This change should be added [hollow-cube@c8358d6](https://github.com/hollow-cube/minestom-ce/commit/c8358d64a176bdf7ef2c51f0cb1f5c05d0fa2543)

Fixed

File: src/main/java/net/minestom/server/damage/DamageType.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import net.minestom.server.utils.NamespaceID;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+
+public sealed interface DamageType extends ProtocolObject, DamageTypes permits DamageTypeImpl {
+    /**
+     * Returns the entity registry.
```
Comment (TheMode) [inline src/main/java/net/minestom/server/damage/DamageType.java:14]: Wrong doc

File: src/main/java/net/minestom/server/damage/DamageTypeImpl.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
+package net.minestom.server.damage;
+
+import net.minestom.server.registry.Registry;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+
+public record DamageTypeImpl(Registry.DamageTypeEntry registry) implements DamageType {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/damage/DamageTypeImpl.java:8]: Should be package-private

File: src/main/java/net/minestom/server/entity/Player.java [hunk 15]
```diff
@@ -248,13 +250,38 @@ public Player(@NotNull UUID uuid, @NotNull String username, @NotNull PlayerConne
     public CompletableFuture<Void> UNSAFE_init(@NotNull Instance spawnInstance) {
         this.dimensionType = spawnInstance.getDimensionType();
 
-        NBTCompound nbt = NBT.Compound(Map.of(
-                "minecraft:chat_type", Messenger.chatRegistry(),
-                "minecraft:dimension_type", MinecraftServer.getDimensionTypeManager().toNBT(),
-                "minecraft:worldgen/biome", MinecraftServer.getBiomeManager().toNBT()));
+        var registry = new HashMap<String, NBT>();
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Player.java:253]: I'd suggest keeping the previous code initiating the immutable compound, but with the `minecraft:damage_type` created just above. As the conversions are a bit unnecessary.

File: src/main/java/net/minestom/server/entity/Player.java [hunk 16]
```diff
@@ -248,13 +250,38 @@ public Player(@NotNull UUID uuid, @NotNull String username, @NotNull PlayerConne
...
+        registry.put("minecraft:chat_type", Messenger.chatRegistry());
+        registry.put("minecraft:dimension_type", MinecraftServer.getDimensionTypeManager().toNBT());
+        registry.put("minecraft:worldgen/biome", MinecraftServer.getBiomeManager().toNBT());
+
+        var damageTypeData = Registry.load(Registry.Resource.DAMAGE_TYPES);
```
Comment (TheMode) [inline src/main/java/net/minestom/server/entity/Player.java:258]: Expensive operation, you should retrieve it from `DamageType.registry()`

File: src/main/java/net/minestom/server/network/packet/server/play/UpdateEnabledFeaturesPacket.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+import net.minestom.server.network.packet.server.ServerPacket;
+import net.minestom.server.network.packet.server.ServerPacketIdentifier;
+import net.minestom.server.utils.NamespaceID;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+
+import static net.minestom.server.network.NetworkBuffer.*;
+
+public record UpdateEnabledFeaturesPacket(Collection<NamespaceID> featureFlags) implements ServerPacket {
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/UpdateEnabledFeaturesPacket.java:13]: Be sure to copy the list to prevent mutation (see other packets taking lists)

File: src/main/java/net/minestom/server/network/packet/server/play/UpdateEnabledFeaturesPacket.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public void write(@NotNull NetworkBuffer writer) {
+        writer.write(VAR_INT, featureFlags.size());
```
Comment (TheMode) [inline src/main/java/net/minestom/server/network/packet/server/play/UpdateEnabledFeaturesPacket.java:16]: There is a dedicated `NetworkBuffer#writeCollection`

**Review by TheMode — CHANGES_REQUESTED** [review]

**Comment by MelonHell** [comment]
I need help with DamageType. I don't understand how registries work

File: src/main/java/net/minestom/server/entity/Player.java [hunk 19]
```diff
@@ -248,10 +250,38 @@ public Player(@NotNull UUID uuid, @NotNull String username, @NotNull PlayerConne
     public CompletableFuture<Void> UNSAFE_init(@NotNull Instance spawnInstance) {
         this.dimensionType = spawnInstance.getDimensionType();
 
+        // TODO replace with something like MinecraftServer.getDamageTypeManager().toNBT()
```
Comment (MelonHell) [inline src/main/java/net/minestom/server/entity/Player.java:253]: I need help here

File: src/main/java/net/minestom/server/damage/DamageType.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    default @NotNull NamespaceID namespace() {
+        return registry().namespace();
+    }
+
+    default double exhaustion() {
+        return registry().exhaustion();
+    }
+
+    default String message_id() {
```
Comment (mworzala) [inline src/main/java/net/minestom/server/damage/DamageType.java:30]: fix case

**Review by mworzala — CHANGES_REQUESTED** [review]

**Comment by MelonHell** [comment]
I figured out what to do with the old DamageType. Since the old DamageType included a type and a source, I decided to rename it Damage and include the amount of damage there.

**Comment by MelonHell** [comment]
I am completely confused. The DamageType registry requires an id that the damage type doesn't have and I don't understand at all what to do and how

**Comment by MelonHell** [comment]
I moved all changes related to Damage to a separate branch
https://github.com/MinestomCommunity/Minestom/pull/3

**Comment by MelonHell** [comment]
this new damage is driving me crazy, I don't know how to properly implement it. I tried to implement looking at EntityTypes as an example, but the damage has no id and is more like biomes

**Comment by MelonHell** [comment]
looks like it's working again
